File: 59_Spiral_Matrix_II/cpp/main.cpp
#incldue <vector>
using namespace std;

vector<vector<int>> generateMatrix(int n) {
	std::vector<std::vector<int>> ret;
	for (int i = 0; i < n; ++i) {
		ret.emplace_back(n);
	}

	int v = 1;
	int q = (n + 1) / 2;
	for (int i = 0; i < q; ++i) {
		int x = i;
		int y = i;
		while (true) {
			ret[x][y] = v++;
			if ((i == q - 1 && n % 2 != 0) || (x == i + 1 && y == i)) {
				break;
			}

			if (x == i && y < n - i - 1) {
				++y;
			}
			else if (y == n - i - 1 && x < n - i - 1) {
				++x;
			}
			else if (x == n - i - 1 && y > i) {
				--y;
			}
			else {
				--x;
			}
		}
	}
	return ret;
}

File: 59_Spiral_Matrix_II/go/main.go
package spiralmmatrix2

func generateMatrix(n int) [][]int {
	ret := make([][]int, n)
	for i := 0; i < n; i++ {
		ret[i] = make([]int, n)
	}

	v := 1
	q := (n + 1) / 2
	for i := 0; i < q; i++ {
		x := i
		y := i

		for {
			ret[x][y] = v
			v++
			if (i == q-1 && n%2 != 0) || (x == i+1 && y == i) {
				break
			}

			switch {
			case x == i && y < n-i-1:
				y++
			case y == n-i-1 && x < n-i-1:
				x++
			case x == n-i-1 && y > i:
				y--
			default:
				x--
			}
		}
	}

	return ret
}
